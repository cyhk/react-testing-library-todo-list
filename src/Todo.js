import React, { Component } from 'react';
// import './App.css';

class Todo extends Component {
  constructor(props) {
    super(props);

    this.state = {
      task: this.props.task,
      edit: false
    }
    this.remove = this.remove.bind(this);
    this.handleEdit = this.handleEdit.bind(this);
    this.handleChange = this.handleChange.bind(this);
    this.saveEdit = this.saveEdit.bind(this);
  }

  remove() {
    this.props.removeFromList(this.props.id);
  }

  handleEdit() {
    this.setState({
      edit: true
    });
  }

  handleChange(evt) {
    this.setState({
      [evt.target.name]: evt.target.value
    })
  }

  saveEdit(evt) {
    evt.preventDefault();
    this.props.editInList(this.props.id, this.state.task);
    this.setState({
      edit: false
    });
  }

  render() {
    if (this.state.edit) {
      return (
        <div>
          <form>
            <input name="task" 
                  onChange={this.handleChange} 
                  data-testid="task-edit-input"
                  value={this.state.task}/>
            <button id="todo-edit-save" onClick={this.saveEdit}>Save</button>
          </form>
        </div>
      );
    } else {
      return (
        <div>
          <div>
            {this.state.task}
          </div>
          <button id="todo-edit" onClick={this.handleEdit}>Edit</button>
          <button id="todo-remove" onClick={this.remove}>Delete</button>
        </div>);
    }

  }
}

export default Todo;
