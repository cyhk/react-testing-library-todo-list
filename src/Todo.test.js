import React from 'react';
import { shallow, mount } from "enzyme";
import toJson from "enzyme-to-json";
import Todo from './Todo';

// smoke tests
it('renders without crashing using shallow', () => {
  shallow(<Todo />);
});

it('renders without crashing using mount', () => {
  mount(<Todo />);
});

// snapshot test
it("matches snapshot", function() {
  let wrapper = shallow(<Todo />);
  let serialized = toJson(wrapper);

  expect(serialized).toMatchSnapshot();
});

// unit tests for functions
it("should run parent handler upon clicking to remove", function(){
  let removeTodo = jest.fn();
  let wrapper = mount(<Todo removeFromList={removeTodo}/>);

  wrapper
    .find('#todo-remove')
    .simulate("click");
  
  expect(removeTodo).toHaveBeenCalled();
});

it("change state upon clicking edit", function(){
  let wrapper = mount(<Todo />);

  wrapper
    .find('#todo-edit')
    .simulate("click")

    expect(wrapper.state().edit).toBe(true);
});

it("change state when input changes", function(){
  let wrapper = mount(<Todo task={"Dig a hole"}/>);

  wrapper
    .find('#todo-edit')
    .simulate("click")

  wrapper
    .find('input[name="task"]')
    .simulate("change", {target: {name: "task", value: "Build a skyscraper"}});

  expect(wrapper.state().task).toBe("Build a skyscraper");
});

it("change state saving an edit", function(){
  const editTodo = jest.fn()
  let wrapper = mount(<Todo task={"Dig another hole"} editInList={editTodo}/>);

  // click edit button
  wrapper
    .find('#todo-edit')
    .simulate("click");

  // edit task
  wrapper
    .find('input[name="task"]')
    .simulate("change", {target: {name: "task", value: "Build a submarine"}});
    
  wrapper
    .find('#todo-edit-save')
    .simulate("click");

  expect(wrapper.state().task).toBe("Build a submarine");
});