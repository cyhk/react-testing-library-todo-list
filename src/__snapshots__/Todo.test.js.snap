// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div>
  <div />
  <button
    id="todo-edit"
    onClick={[Function]}
  >
    Edit
  </button>
  <button
    id="todo-remove"
    onClick={[Function]}
  >
    uX
  </button>
</div>
`;
